README.md
Implementation for Stereo DSO from TUM.

Stereo completed. Now adding IMU, the main IMU code is in CoarseTracker.

The results of mine compared to ORB_SLAM2 in the 11 KITTI sequences are in the folder `results`.

I removed loop closure and relocalization in the ORB_SLAM2, so that the ORB_SLAM2 is much like Stereo DSO for it's an visual odometry and only 2d-2d tracking.

Some results are in this readme file, section 6.

# DSO: Direct Sparse Odometry

For more information see
[https://vision.in.tum.de/dso](https://vision.in.tum.de/dso)

### 1. Related Papers
* **Direct Sparse Odometry**, *J. Engel, V. Koltun, D. Cremers*, In arXiv:1607.02565, 2016
* **A Photometrically Calibrated Benchmark For Monocular Visual Odometry**, *J. Engel, V. Usenko, D. Cremers*, In arXiv:1607.02555, 2016

Get some datasets from [https://vision.in.tum.de/mono-dataset](https://vision.in.tum.de/mono-dataset) .

### 2. Installation

	git clone https://github.com/JakobEngel/dso.git

#### 2.1 Required Dependencies

##### suitesparse and eigen3 (required).
Required. Install with

		sudo apt-get install libsuitesparse-dev libeigen3-dev libboost-all-dev



#### 2.2 Optional Dependencies

##### OpenCV (highly recommended).
Used to read / write / display images.
OpenCV is **only** used in `IOWrapper/OpenCV/*`. Without OpenCV, respective 
dummy functions from `IOWrapper/*_dummy.cpp` will be compiled into the library, which do nothing.
The main binary will not be created, since it is useless if it can't read the datasets from disk.
Feel free to implement your own version of these functions with your prefered library, 
if you want to stay away from OpenCV.

Install with

	sudo apt-get install libopencv-dev


##### Pangolin (highly recommended).
Used for 3D visualization & the GUI.
Pangolin is **only** used in `IOWrapper/Pangolin/*`. You can compile without Pangolin, 
however then there is not going to be any visualization / GUI capability. 
Feel free to implement your own version of `Output3DWrapper` with your preferred library, 
and use it instead of `PangolinDSOViewer`

Install from [https://github.com/stevenlovegrove/Pangolin](https://github.com/stevenlovegrove/Pangolin)


##### ziplib (recommended).
Used to read datasets with images as .zip, as e.g. in the TUM monoVO dataset. 
You can compile without this, however then you can only read images directly (i.e., have 
to unzip the dataset image archives before loading them).

	sudo apt-get install zlib1g-dev
	cd dso/thirdparty
	tar -zxvf libzip-1.1.1.tar.gz
	cd libzip-1.1.1/
	./configure
	make
	sudo make install
	sudo cp lib/zipconf.h /usr/local/include/zipconf.h   # (no idea why that is needed).

##### sse2neon (required for ARM builds).
After cloning, just run `git submodule update --init` to include this.  It translates Intel-native SSE functions to ARM-native NEON functions during the compilation process.

#### 2.3 Build

		cd dso
		mkdir build
		cd build
		cmake ..
		make -j
	
this will compile a library `libdso.a`, which can be linked from external projects. 
It will also build a binary `dso_dataset`, to run DSO on datasets. However, for this
OpenCV and Pangolin need to be installed.






### 3 Usage
Run on a dataset from [https://vision.in.tum.de/mono-dataset](https://vision.in.tum.de/mono-dataset) using

		bin/dso_dataset \
			files=XXXXX/sequence_XX/images.zip \
			calib=XXXXX/sequence_XX/camera.txt \
			gamma=XXXXX/sequence_XX/pcalib.txt \
			vignette=XXXXX/sequence_XX/vignette.png \
			preset=0 \
			mode=0

See [https://github.com/JakobEngel/dso_ros](https://github.com/JakobEngel/dso_ros) for a minimal example on
how the library can be used from another project. It should be straight forward to implement extentions for 
other camera drivers, to use DSO interactively without ROS.



#### 3.1 Dataset Format.
The format assumed is that of [https://vision.in.tum.de/mono-dataset](https://vision.in.tum.de/mono-dataset).
However, it should be easy to adapt it to your needs, if required. The binary is run with:

- `files=XXX` where XXX is either a folder or .zip archive containing images. They are sorted *alphabetically*. for .zip to work, need to comiple with ziplib support.

- `gamma=XXX` where XXX is a gamma calibration file, containing a single row with 256 values, mapping [0..255] to the respective irradiance value, i.e. containing the *discretized inverse response function*. See TUM monoVO dataset for an example.

- `vignette=XXX` where XXX is a monochrome 16bit or 8bit image containing the vignette as pixelwise attenuation factors. See TUM monoVO dataset for an example.

- `calib=XXX` where XXX is a geometric camera calibration file. See below.



##### Geometric Calibration File.


###### Calibration File for Pre-Rectified Images

    Pinhole fx fy cx cy 0
    in_width in_height
    "crop" / "full" / "none" / "fx fy cx cy 0"
    out_width out_height

###### Calibration File for FOV camera model:

    FOV fx fy cx cy omega
    in_width in_height
    "crop" / "full" / "fx fy cx cy 0"
    out_width out_height


###### Calibration File for Radio-Tangential camera model

    RadTan fx fy cx cy k1 k2 r1 r2
    in_width in_height
    "crop" / "full" / "fx fy cx cy 0"
    out_width out_height


###### Calibration File for Equidistant camera model

    EquiDistant fx fy cx cy k1 k2 r1 r2
    in_width in_height
    "crop" / "full" / "fx fy cx cy 0"
    out_width out_height


(note: for backwards-compatibility, "Pinhole", "FOV" and "RadTan" can be omitted). See the respective
`::distortCoordinates` implementation in  `Undistorter.cpp` for the exact corresponding projection function.
Furthermore, it should be straight-forward to implement other camera models.


**Explanation:**
 Across all models `fx fy cx cy` denotes the focal length / principal point **relative to the image width / height**, 
i.e., DSO computes the camera matrix `K` as

		K(0,0) = width * fx
		K(1,1) = height * fy
		K(0,2) = width * cx - 0.5
		K(1,2) = height * cy - 0.5
For backwards-compatibility, if the given `cx` and `cy` are larger than 1, DSO assumes all four parameters to directly be the entries of K, 
and ommits the above computation. 


**That strange "0.5" offset:**
 Internally, DSO uses the convention that the pixel at integer position (1,1) in the image, i.e. the pixel in the second row and second column,
contains the integral over the continuous image function from (0.5,0.5) to (1.5,1.5), i.e., approximates a "point-sample" of the 
continuous image functions at (1.0, 1.0).
In turn, there seems to be no unifying convention across calibration toolboxes whether the pixel at integer position (1,1)
contains the integral over (0.5,0.5) to (1.5,1.5), or the integral over (1,1) to (2,2). The above conversion assumes that 
the given calibration in the calibration file uses the latter convention, and thus applies the -0.5 correction.
Note that this also is taken into account when creating the scale-pyramid (see `globalCalib.cpp`).


**Rectification modes:**
 For image rectification, DSO either supports rectification to a user-defined pinhole model (`fx fy cx cy 0`),
or has an option to automatically crop the image to the maximal rectangular, well-defined region (`crop`).
`full` will preserve the full original field of view and is mainly meant for debugging - it will create black 
borders in undefined image regions, which DSO does NOT ignore (i.e., this option will generate additional 
outliers along those borders, and corrupt the scale-pyramid).




#### 3.2 Commandline Options
there are many command line options available, see `main_dso_pangolin.cpp`. some examples include
- `mode=X`: 
    -  `mode=0` use iff a photometric calibration exists (e.g. TUM monoVO dataset). 
    -  `mode=1` use iff NO photometric calibration exists (e.g. ETH EuRoC MAV dataset). 
    -  `mode=2` use iff images are not photometrically distorted (e.g. synthetic datasets).

- `preset=X`
    - `preset=0`: default settings (2k pts etc.), not enforcing real-time execution
    - `preset=1`: default settings (2k pts etc.), enforcing 1x real-time execution
    - `preset=2`: fast settings (800 pts etc.), not enforcing real-time execution. WARNING: overwrites image resolution with 424 x 320.
    - `preset=3`: fast settings (800 pts etc.), enforcing 5x real-time execution. WARNING: overwrites image resolution with 424 x 320.

- `nolog=1`: disable logging of eigenvalues etc. (good for performance)
- `reverse=1`: play sequence in reverse
- `nogui=1`: disable gui (good for performance)
- `nomt=1`: single-threaded execution
- `prefetch=1`: load into memory & rectify all images before running DSO.
- `start=X`: start at frame X
- `end=X`: end at frame X
- `speed=X`: force execution at X times real-time speed (0 = not enforcing real-time)
- `save=1`: save lots of images for video creation
- `quiet=1`: disable most console output (good for performance)
- `sampleoutput=1`: register a "SampleOutputWrapper", printing some sample output data to the commandline. meant as example.



#### 3.3 Runtime Options
Some parameters can be reconfigured from the Pangolin GUI at runtime. Feel free to add more.


#### 3.4 Accessing Data.
The easiest way to access the Data (poses, pointclouds, etc.) computed by DSO (in real-time)
is to create your own `Output3DWrapper`, and add it to the system, i.e., to `FullSystem.outputWrapper`.
The respective member functions will be called on various occations (e.g., when a new KF is created, 
when a new frame is tracked, etc.), exposing the relevant data.

See `IOWrapper/Output3DWrapper.h` for a description of the different callbacks available,
and some basic notes on where to find which data in the used classes.
See `IOWrapper/OutputWrapper/SampleOutputWrapper.h` for an example implementation, which just prints
some example data to the commandline (use the options `sampleoutput=1 quiet=1` to see the result).

Note that these callbacks block the respective DSO thread, thus expensive computations should not
be performed in the callbacks, a better practice is to just copy over / publish / output the data you need.

Per default, `dso_dataset` writes all keyframe poses to a file `result.txt` at the end of a sequence,
using the TUM RGB-D / TUM monoVO format ([timestamp x y z qx qy qz qw] of the cameraToWorld transformation).



#### 3.5 Notes
- the initializer is very slow, and does not work very reliably. Maybe replace by your own way to get an initialization.
- see [https://github.com/JakobEngel/dso_ros](https://github.com/JakobEngel/dso_ros) for a minimal example project on how to use the library with your own input / output procedures.
- see `settings.cpp` for a LOT of settings parameters. Most of which you shouldn't touch.
- `setGlobalCalib(...)` needs to be called once before anything is initialized, and globally sets the camera intrinsics and video resolution for convenience. probably not the most portable way of doing this though.




### 4 General Notes for Good Results

#### Accurate Geometric Calibration
- Please have a look at Chapter 4.3 from the DSO paper, in particular Figure 20 (Geometric Noise). Direct approaches suffer a LOT from bad geometric calibrations: Geometric distortions of 1.5 pixel already reduce the accuracy by factor 10.


history.txt
    1  ls
    2  exit
    3  ls
    4  exit
    5  sudo apt-get install python3.7
    6  sudo apt install python3.7
    7  sudo apt update
    8  exit
    9  sudo apt-get update
   10  exit
   11  pip3 install tensorflow-gpu
   12  pip --version
   13  pip3 install --upgrade pip3
   14  pip3 install --upgrade pip
   15  pip --version
   16  which pip
   17  which pip3
   18  cd /home/jingetu/.local
   19  cd bin
   20  ls
   21  ls -lh pip*
   22  sudo pip install keras
   23  sudo pip3 install keras
   24  pip3 install keras
   25  sudo pip3 install keras
   26  exit
   27  ./cifar10.py
   28  vim cifar10.py
   29  sudo apt-get install vim
   30  ls
   31  vim cifar10.py
   32  chmod +x cifar10.py
   33  ./cifar10.py
   34  ls
   35  s
   36  vim cifar10.py
   37  ls
   38  exit
   39  ls
   40  cd Desktop
   41  ls
   42  cd work
   43  ls
   44  cd ..
   45  ls
   46  cd Downloads
   47  ls
   48  cd keras-resnet
   49  ls
   50  code .
   51  ls -lh
   52  ls
   53  mkdir first_round
   54  ls
   55  mv *.hdf5 first_round
   56  ls
   57  cd ..
   58  ls
   59  cd keras-resnet
   60  ls
   61  python3 cifar10.py
   62  sl
   63  ls
   64  cd /media
   65  ls
   66  cd jingetu
   67  ls
   68  cd ..
   69  ls
   70  cd xin
   71  ls
   72  cd jingetu
   73  ls
   74  cd ..
   75  cd jingetu
   76  ls
   77  cd ..
   78  cd jingetu
   79  ls
   80  cd UBUNTU\ 16_0
   81  ls
   82  cd UBUNTU\ 16_0
   83  ls
   84  ls -lh
   85  ln -s xinjiajuan 新加卷
   86  sudo ln -s xinjiajuan 新加卷
   87  cd 新加卷
   88  ls
   89  cd youtube
   90  ls
   91  cd cs231n
   92  ls
   93  ./test_download_v1.1.py
   94  pip --version
   95  pip install youtube-dl
   96  exit
   97  bash
   98  sudo pip install youtube-dl
   99  pip install youtube-dl --user
  100  ls
  101  exit
  102  sudo apt install jq
  103  ls
  104  cd Desktop
  105  ls
  106  cd work
  107  ls
  108  cd ..
  109  ls
  110  cd /media
  111  ls
  112  cd jige
  113  cd jingetu
  114  ls
  115  cd 新加卷
  116  ls
  117  cd youtube
  118  ls
  119  cd cs231n
  120  lks
  121  ls
  122  ./test_download_v1.1.py
  123  sudo apt-get install vlc
  124  nvidia-smi
  125  ls
  126  cd Desktop
  127  ks
  128  ls
  129  mkdir work
  130  ls
  131  rm -rf work
  132  ls
  133  mkdir workspace
  134  cd workspace
  135  ls
  136  mkdir test
  137  ls
  138  cd test
  139  ls
  140  vim main.cpp
  141  g++ main.cpp
  142  vim main.cpp
  143  g++ main.cpp
  144  ./a.out
  145  g++ main.cpp
  146  vim main.cpp
  147  cat /etc/issue
  148  sudo apt-get install fcitx-googlepinyin
  149  sudo apt-get update
  150  sudo apt-get install fcitx-googlepinyin
  151  sudo rm /var/lib/dpkg/lock-frontend
  152  sudo apt-get install fcitx-googlepinyin
  153  sudo rm /var/lib/dpkg/lock
  154  sudo apt-get install fcitx-googlepinyin
  155  sudo dpkg --configure -a
  156  ps -ef | grep dpkg
  157  kill -9 27284
  158  sudo kill -9 27284
  159  sudo dpkg --configure -a
  160  sudo apt-get install fcitx-googlepinyin
  161  sudo apt install -y fcitx-bin
  162  sudo apt autoremove
  163  ls
  164  sudo apt-get install ./sogoupinyin_2.2.0.0108_amd64.deb
  165  sudo apt-get remove ./sogoupinyin_2.2.0.0108_amd64.deb
  166  sudo apt-get remove sogoupinyin
  167  sudo apt-get remove ./sogoupinyin_2.2.0.0108_amd64.deb
  168  sudo apt-get install ./sogoupinyin_2.2.0.0108_amd64.deb
  169  sudo apt-get remove sogoupinyin
  170  sudo apt-get install ./sogoupinyin_2.2.0.0108_amd64.deb
  171  sudo apt-get remove fcitx
  172  sudo apt-get remove fcitx-bin
  173  sudo apt install -y fcitx-bin
  174  sudo apt install -y fcitx-googlepinyin
  175  wget "http://cdn2.ime.sogou.com/dl/index/1524572264/sogoupinyin_2.2.0.0108_amd64.deb?st=qC_O2p5443g1a2TJR_rSdA&e=1533163019&fn=sogoupinyin_2.2.0.0108_amd64.deb" -O sogoupinyin.deb && \\n    yes | sudo gdebi sogoupinyin.deb && \\n    rm sogoupinyin.deb
  176  sudo apt-get install gdebi
  177  wget "http://cdn2.ime.sogou.com/dl/index/1524572264/sogoupinyin_2.2.0.0108_amd64.deb?st=qC_O2p5443g1a2TJR_rSdA&e=1533163019&fn=sogoupinyin_2.2.0.0108_amd64.deb" -O sogoupinyin.deb && \\n    yes | sudo gdebi sogoupinyin.deb && \\n    rm sogoupinyin.deb
  178  sudo apt-get remove sogoupinyin
  179  sudo dpkg
  180  sudo dpkg -i ./sogoupinyin_2.2.0.0108_amd64.deb
  181  sudo apt-get remove sogoupinyin
  182  sudo apt-get autoremove
  183  cd Downloads
  184  ls
  185  sudo dpkg -i ./sogoupinyin_2.2.0.0108_amd64.deb
  186  sudo apt-get install fcitx
  187  sudo apt-get -f install
  188  cat /usr/share/glib-2.0/schemas/50_sogoupinyin.gschema.override
  189  vim cat /usr/share/glib-2.0/schemas/50_sogoupinyin.gschema.override
  190  sudo apt-get install fcitx
  191  sudo nvidia-smi
  192  sudo add-apt-repository ppa:gnome3-team/gnome3
  193  sudo apt-get update && sudo apt-get upgrade
  194  sudo add-apt-repository ppa:gnome3-team/gnome3
  195  sudo apt-get update && sudo apt-get upgrade
  196  sudo apt-get install gnome-shell ubuntu-gnome-desktop
  197  gnome-shell
  198  ubuntu-gnome-desktop
  199  gnome-shell --help
  200  gnome-shell --replace gdm3
  201  gnome-shell --replace
  202  exit
  203  sudo apt-get remove sogoupinyin
  204  sudo apt-get remove unity lightdm ubuntu-desktop
  205  ls
  206  sudo dpkg -i sogoupinyin_2.2.0.0108_amd64.deb
  207  cat /usr/share/glib-2.0/schemas/50_sogoupinyin.gschema.override
  208  sudo glib-compile-schemas /usr/share/glib-2.0/schemas/
  209  sudo apt install fcitx fcitx-googlepinyin fcitx-table-wbpy fcitx-pinyin fcitx-sunpinyin
  210  sudo remove ibus
  211  sudo apt remove ibus
  212  sudo dpkg -i ~/Downloads/sogoupinyin*.deb; sudo apt -f install
  213  sudo apt-get remove sogoupinyin

src/util/settings.cpp
/**
* This file is part of DSO.
* 
* Copyright 2016 Technical University of Munich and Intel.
* Developed by Jakob Engel <engelj at in dot tum dot de>,
* for more information see <http://vision.in.tum.de/dso>.
* If you use this code, please cite the respective publications as
* listed on the above website.
*
* DSO is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* DSO is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with DSO. If not, see <http://www.gnu.org/licenses/>.
*/



#include "util/settings.h"
#include <boost/bind.hpp>


namespace dso {
  int pyrLevelsUsed = PYR_LEVELS;

  bool setting_PREnofixKF = false;
/* Parameters controlling when KF's are taken */
  float setting_keyframesPerSecond = 0;   // if !=0, takes a fixed number of KF per second.
  bool setting_realTimeMaxKF = false;   // if true, takes as many KF's as possible (will break the system if the camera stays stationary)
  float setting_maxShiftWeightT = 0.04f * (752 + 480);
  float setting_maxShiftWeightR = 0.0f * (752 + 480);
  float setting_maxShiftWeightRT = 0.02f * (752 + 480);
  float setting_kfGlobalWeight = 1;   // general weight on threshold, the larger the more KF's are taken (e.g., 2 = double the amount of KF's).
  float setting_maxAffineWeight = 2;
  float setting_imuResidualWeight = 0.1;
  float setting_imuResidualWeightInv = 1 / setting_imuResidualWeight;
  float setting_imuResidualWeightSquare = setting_imuResidualWeight * setting_imuResidualWeight;
  float setting_imuResidualWeightSquareInv = 1 / setting_imuResidualWeightSquare;


/* initial hessian values to fix unobservable dimensions / priors on affine lighting parameters.
 */
  float setting_idepthFixPrior = 50 * 50;
  float setting_idepthFixPriorMargFac = 600 * 600;
  float setting_initialRotPrior = 1e11;
  float setting_initialTransPrior = 1e10;
  float setting_initialAffBPrior = 1e14;
  float setting_initialAffAPrior = 1e14;
  float setting_initialCalibHessian = 5e9;


/* some modes for solving the resulting linear system (e.g. orthogonalize wrt. unobservable dimensions) */
  int setting_solverMode =
      SOLVER_FIX_LAMBDA | SOLVER_ORTHOGONALIZE_X_LATER; //- original SOLVER_FIX_LAMBDA | SOLVER_ORTHOGONALIZE_X_LATER
  double setting_solverModeDelta = 0.00001;
  bool setting_forceAceptStep = true;


/* some thresholds on when to activate / marginalize points */
  float setting_minIdepthH_act = 100;
  float setting_minIdepthH_marg = 50;


  float setting_desiredImmatureDensity = 1500; // immature points per frame
  float setting_desiredPointDensity = 2000; // aimed total points in the active window.
  float setting_minPointsRemaining = 0.05;  // marg a frame if less than X% points remain.
  float setting_maxLogAffFacInWindow = 0.7; // marg a frame if factor between intensities to current frame is larger than 1/X or X.


  int setting_minFrames = 5; // min frames in window.
  int setting_maxFrames = 7; // max frames in window.
  int setting_minFrameAge = 1;
  int setting_maxOptIterations = 6; // max GN iterations.
  int setting_minOptIterations = 1; // min GN iterations.
  float setting_thOptIterations = 1.2; // factor on break threshold for GN iteration (larger = break earlier)





/* Outlier Threshold on photometric energy */
  float setting_outlierTH = 12 * 12;          // higher -> less strict
  float setting_outlierTHSumComponent = 50 * 50;    // higher -> less strong gradient-based reweighting .




  int setting_pattern = 8;            // point pattern used. DISABLED.
  float setting_margWeightFac =
      0.5 * 0.5;          // factor on hessian when marginalizing, to account for inaccurate linearization points.


/* when to re-track a frame */
  float setting_reTrackThreshold = 1.5; // (larger = re-track more often)



/* require some minimum number of residuals for a point to become valid */
  int setting_minGoodActiveResForMarg = 3;
  int setting_minGoodResForMarg = 4;


// 0 = nothing.
// 1 = apply inv. response.
// 2 = apply inv. response & remove V.
  int setting_photometricCalibration = 2;
  bool setting_useExposure = true;
  float setting_affineOptModeA = 1e12; //-1: fix. >=0: optimize (with prior, if > 0).
  float setting_affineOptModeB = 1e8; //-1: fix. >=0: optimize (with prior, if > 0).

  int setting_gammaWeightsPixelSelect = 1; // 1 = use original intensity for pixel selection; 0 = use gamma-corrected intensity.




  float setting_huberTH = 9; // Huber Threshold

  float setting_staticLambda = 0.5; //- For temporal stereo and static stereo combination.

  float setting_acceptStaticDepthFactor = 200;
  float setting_temporal_imu_data_overlap = 0.02;

// parameters controlling adaptive energy threshold computation.
  float setting_frameEnergyTHConstWeight = 0.5;
  float setting_frameEnergyTHN = 0.7f;
  float setting_frameEnergyTHFacMedian = 1.5;
  float setting_overallEnergyTHWeight = 1;
  float setting_coarseCutoffTH = 20;


// parameters controlling pixel selection
  float setting_minGradHistCut = 0.5;
  float setting_minGradHistAdd = 7;
  float setting_gradDownweightPerLevel = 0.75;
  bool setting_selectDirectionDistribution = true;


/* settings controling initial immature point tracking */
  float setting_maxPixSearch = 0.027; // max length of the ep. line segment searched during immature point tracking. relative to image resolution.
  float setting_minTraceQuality = 3;
  int setting_minTraceTestRadius = 2;
  int setting_GNItsOnPointActivation = 3;
  float setting_trace_stepsize = 1.0;        // stepsize for initial discrete search.
  int setting_trace_GNIterations = 3;        // max # GN iterations
  float setting_trace_GNThreshold = 0.1;        // GN stop after this stepsize.
  float setting_trace_extraSlackOnTH = 1.2;      // for energy-based outlier check, be slightly more relaxed by this factor.
  float setting_trace_slackInterval = 1.5;      // if pixel-interval is smaller than this, leave it be.
  float setting_trace_minImprovementFactor = 2;    // if pixel-interval is smaller than this, leave it be.




// for benchmarking different undistortion settings
  float benchmarkSetting_fxfyfac = 0;
  int benchmarkSetting_width = 0;
  int benchmarkSetting_height = 0;
  float benchmark_varNoise = 0;
  float benchmark_varBlurNoise = 0;
  float benchmark_initializerSlackFactor = 1;
  int benchmark_noiseGridsize = 3;


  float freeDebugParam1 = 1;
  float freeDebugParam2 = 1;
  float freeDebugParam3 = 1;
  float freeDebugParam4 = 1;
  float freeDebugParam5 = 1;


  bool disableReconfigure = false;
  bool debugSaveImages = false;
  bool multiThreading = false; //- original true
  bool disableAllDisplay = false;
  bool setting_onlyLogKFPoses = false;
  bool setting_logStuff = true;


  bool goStepByStep = false;


  bool setting_render_displayCoarseTrackingFull = false;
  bool setting_render_renderWindowFrames = true;
  bool setting_render_plotTrackingFull = false;
  bool setting_render_display3D = true;
  bool setting_render_displayResidual = true;
  bool setting_render_displayVideo = true;
  bool setting_render_displayDepth = true;

  bool setting_fullResetRequested = false;

  bool setting_debugout_runquiet = false;

  int sparsityFactor = 5;  // not actually a setting, only some legacy stuff for coarse initializer.


  void handleKey(char k) {
    char kkk = k;
    switch (kkk) {
      case 'd':
      case 'D':
        freeDebugParam5 = ((int) (freeDebugParam5 + 1)) % 10;
        printf("new freeDebugParam5: %f!\n", freeDebugParam5);
        break;
      case 's':
      case 'S':
        freeDebugParam5 = ((int) (freeDebugParam5 - 1 + 10)) % 10;
        printf("new freeDebugParam5: %f!\n", freeDebugParam5);
        break;
    }

  }


  int staticPattern[10][40][2] = {
      {{0,  0},  {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},  // .
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}},

      {{0,  -1}, {-1,   0},    {0,    0},    {1,    0},    {0,    1},    {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},  // +
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}},

      {{-1, -1}, {1,    1},    {0,    0},    {-1,   1},    {1,    -1},   {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},  // x
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}},

      {{-1, -1}, {-1,   0},    {-1,   1},    {-1,   0},    {0,    0},    {0,    1},    {1,    -1},   {1,    0},    {1,    1},    {-100, -100},  // full-tight
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}},

      {{0,  -2}, {-1,   -1},   {1,    -1},   {-2,   0},    {0,    0},    {2,    0},    {-1,   1},    {1,    1},    {0,    2},    {-100, -100},  // full-spread-9
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}},

      {{0,  -2}, {-1,   -1},   {1,    -1},   {-2,   0},    {0,    0},    {2,    0},    {-1,   1},    {1,    1},    {0,    2},    {-2,   -2},   // full-spread-13
          {-2,   2},    {2,    -2},   {2,    2},    {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}},

      {{-2, -2}, {-2,   -1},   {-2,   -0},   {-2,   1},    {-2,   2},    {-1,   -2},   {-1,   -1},   {-1,   -0},   {-1,   1},    {-1,   2},                    // full-25
          {-0,   -2},   {-0,   -1},   {-0,   -0},   {-0,   1},    {-0,   2},    {+1,   -2},   {+1,   -1},   {+1,   -0},   {+1,   1},    {+1,   2},
          {+2,   -2},   {+2,   -1},   {+2,   -0},   {+2,   1},    {+2,   2},    {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}},

      {{0,  -2}, {-1,   -1},   {1,    -1},   {-2,   0},    {0,    0},    {2,    0},    {-1,   1},    {1,    1},    {0,    2},    {-2,   -2},   // full-spread-21
          {-2,   2},    {2,    -2},   {2,    2},    {-3,   -1},   {-3,   1},    {3,    -1},   {3,    1},    {1,    -3},   {-1,   -3},   {1,    3},
          {-1,   3},    {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}},

      {{0,  -2}, {-1,   -1},   {1,    -1},   {-2,   0},    {0,    0},    {2,    0},    {-1,   1},    {0,    2},    {-100, -100}, {-100, -100},  // 8 for SSE efficiency
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100},
          {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}, {-100, -100}},

      {{-4, -4}, {-4,   -2},   {-4,   -0},   {-4,   2},    {-4,   4},    {-2,   -4},   {-2,   -2},   {-2,   -0},   {-2,   2},    {-2,   4},                    // full-45-SPREAD
          {-0,   -4},   {-0,   -2},   {-0,   -0},   {-0,   2},    {-0,   4},    {+2,   -4},   {+2,   -2},   {+2,   -0},   {+2,   2},    {+2,   4},
          {+4,   -4},   {+4,   -2},   {+4,   -0},   {+4,   2},    {+4,   4},    {-200, -200}, {-200, -200}, {-200, -200}, {-200, -200}, {-200, -200},
          {-200, -200}, {-200, -200}, {-200, -200}, {-200, -200}, {-200, -200}, {-200, -200}, {-200, -200}, {-200, -200}, {-200, -200}, {-200, -200}},
  };

  int staticNCCPattern35[15][2] = {
      {-2, -1},
      {-1, -1},
      {0,  -1},
      {1,  -1},
      {2,  -1},
      {-2, 0},
      {-1, 0},
      {0,  0},
      {1,  0},
      {2,  0},
      {-2, 1},
      {-1, 1},
      {0,  1},
      {1,  1},
      {2,  1},

  };

  int staticPatternNum[10] = {
      1,
      5,
      5,
      9,
      9,
      13,
      25,
      21,
      8,
      25
  };

  int staticPatternPadding[10] = {
      1,
      1,
      1,
      1,
      2,
      2,
      2,
      3,
      2,
      4
  };


}
